src/tools/tool_custom.rs

```diff
@@ -0,0 +1,225 @@
+use std::any::Any;
+use std::collections::HashMap;
+use std::sync::Arc;
+use serde_json::Value;
+use tokio::io::BufReader;
+use async_trait::async_trait;
+use regex::Regex;
+use tokio::sync::{Mutex as AMutex, RwLock as ARwLock};
+use tracing::{info, warn};
+use tokio::process::{Command, Child, ChildStdout, ChildStderr};
+use tokio::time::{timeout, Duration};
+use std::process::Stdio;
+
+use crate::at_commands::at_commands::AtCommandsContext;
+use crate::tools::tools_description::{AtParamDict, Tool};
+use crate::call_validation::{ChatMessage, ChatContent, ContextEnum};
+use crate::global_context::GlobalContext;
+use crate::integrations::process_io_utils::{kill_process_and_children, read_until_token_or_timeout};
+use crate::integrations::sessions::IntegrationSession;
+
+
+pub struct ToolCustom {
+    pub name: String,
+    #[allow(dead_code)]
+    pub parameters: Vec<AtParamDict>,
+    pub parameters_required: Vec<String>,
+    pub command: String,
+    pub runs_in_background: bool,
+    pub runs_in_background_false_timeout: usize,
+}
+
+pub struct ToolSession {
+    process: Child,
+    #[allow(dead_code)]
+    stdout: BufReader<ChildStdout>,
+    #[allow(dead_code)]
+    stderr: BufReader<ChildStderr>,
+}
+
+impl IntegrationSession for ToolSession {
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+    fn is_expired(&self) -> bool { false }
+}
+
+fn replace_magics_from_command(
+    command: &str,
+    args: &HashMap<String, Value>,
+    required_params: &Vec<String>
+) -> Result<String, String> {
+    let args_str: HashMap<String, String> = args.iter()
+        .filter_map(|(k, v)| v.as_str().map(|s| (k.clone(), s.to_string())))
+        .collect();
+
+    let mut command_clone = command.to_string();
+    for (key, value) in args_str {
+        let pattern = format!("%{}%", key);
+        let re = Regex::new(&regex::escape(&pattern)).unwrap();
+        command_clone = re.replace_all(&command_clone, value.as_str()).to_string();
```

Comment:
don't need regex

Reply:
I need to replace all occurencies, therefore I use regex